gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/DiskStoreImpl.java

```diff
@@ -5260,4 +5327,110 @@ else if (memoryState.isEviction()) {
       resetAsyncQueueCapacity();
     }
   }
+
+  public void acquireDiskStoreReadLock() {
+    try {
+      this.diskStoreLock.readLock().acquire();
+    } catch(InterruptedException ie) {
+      this.stopper.checkCancelInProgress(ie);
+      Thread.currentThread().interrupt();
+      throw new InternalGemFireException(ie);
+    }
+  }
+
+  public void releaseDiskStoreReadLock() {
+    this.diskStoreLock.readLock().release();
+  }
+
+  public void acquireDiskStoreWriteLock() {
+    try {
+      this.diskStoreLock.writeLock().acquire();
+    } catch (InterruptedException ie) {
+      this.stopper.checkCancelInProgress(ie);
+      Thread.currentThread().interrupt();
+      throw new InternalGemFireException(ie);
+    }
+  }
+
+  public void releaseDiskStoreWriteLock() {
+    this.diskStoreLock.writeLock().release();
+  }
+
+  void shutdownDiskStoreAndAffiliatedRegions(DiskAccessException dae) {
+    final ThreadGroup exceptionHandlingGroup = LogWriterImpl.createThreadGroup(
+      "Disk Store Exception Handling Group", cache.getLoggerI18n());
+
+    // Shutdown the regions and bridge servers in another thread, to make sure
+    // that we don't cause a deadlock because this thread is holding some lock
+    Thread thread = new Thread(exceptionHandlingGroup,
+      "Disk store exception handler") {
+      @Override
+      public void run() {
+
+        // wait in a loop for Transactions to be empty
+        TXManagerImpl txManager = DiskStoreImpl.this.cache.getTxManager();
+        // wait for 5 minutes for tx to be empty , then just close?
+        int count = 0;
+        DiskStoreImpl.this.acquireDiskStoreWriteLock();
+        while(txManager.hasHostedTransactions()) {
```

Comment:
Hosted transactions might keep on building. Are we making sure somewhere that new txns are not being started. If yes, where?

Reply:
Or are you counting on the write preference nature of the lock?
This might end up in an infinite loop. Isn't it?

Reply:
The hing which I hope will prevent this:
The writer pref lock as you said, is helpful but not too that extent because transaction threads would not acquire this lock. The non tx threads will be positively affected by this lock.  The Tx thread if starting a transaction or if in progress will get an exception ,causing them to rollback. I am counting on that , that after some time user will stop executing transactions. Also the loop waits for around 10 minutes for all tx to end. If it still is not able to , then the thread just proceeds to close the disk store.

Reply:
Ok. Yes as you explained that the tx will get exception when they try to write into the disk store. Its just that theoretically hostedTransactions can keep on increasing too. Looks like should be ok.